### PR TITLE
Feature/pep xml prophets

### PR DESCRIPTION
**Pull request by hroest**
- improve pepXML parsing
- add support for the `search_score_summary` tag
- add support to store interProphet *and* peptideProphet results
- support for additional attributes in `spectrum_query`
- support to write out pepxml attributes

### CODE REVIEW

File: src/openms/source/FORMAT/PepXMLFile.cpp [hunk 1]
```diff
@@ -71,6 +74,8 @@ namespace OpenMS
 
   void PepXMLFile::store(const String& filename, std::vector<ProteinIdentification>& protein_ids, std::vector<PeptideIdentification>& peptide_ids, const String& mz_file, const String& mz_name, bool peptideprophet_analyzed)
   {
+    bool use_original_spectrum_name = false;
```
Comment (hroest) [inline src/openms/source/FORMAT/PepXMLFile.cpp]: @hendrikweisser if we turn this on (or at least make it an option) it would allow us to retain the original `<spectrum_query spectrum="hroest_K120718_SM_OGE10_010_IDA.02552.02552.2"` information when writing out pep.xml -> personally I would prefer that but it might break some existing code which basically uses `PepXMLFile_test.00000.00000.3` which to me seems inferior

is there some code that relies on the OpenMS `0000.0000` version?

File: src/openms/source/FORMAT/PepXMLFile.cpp [hunk 2]
```diff
@@ -804,22 +913,81 @@ namespace OpenMS
       // may appear to be "out of bounds" - see NOTE above:
       String identifier = current_proteins_[min(UInt(current_proteins_.size()), search_id_) - 1]->getIdentifier();
       current_peptide_.setIdentifier(identifier);
+
+      // set optional attributes
+      if (!native_spectrum_name_.empty()) 
+      {
+        current_peptide_.setMetaValue("pepxml_spectrum_name", native_spectrum_name_);
```
Comment (hroest) [inline src/openms/source/FORMAT/PepXMLFile.cpp]: this leads to a few testfailues and a `UserParam` to be present in every identifcation in the idXML file. Maybe we can discuss whether we really want this or not, we might also skip it and not keep that information around

**Comment by hroest** [comment]
I believe I have solved the issue with the native name by letting the user decide and adding a flag

**Comment by hendrikweisser** [comment]
> I believe I have solved the issue with the native name by letting the user decide and adding a flag

I didn't work on any of the code for writing, only reading. My understanding is that the "old" code for writing pepXML was only a minimalistic hack in order to produce input files for ProteinProphet. So I don't see why a flag should be necessary - your support for the native name looks like a clear improvement (provided that TPP tools can still deal with the output).

**Comment by timosachsenberg** [comment]
so should we merge as is or test against TPP and remove the flag?

**Comment by grosenberger** [comment]
There might be a small incompatibilty with one of the pyOpenMS unit tests:
```
<Non-fatal error while loading './pyOpenMS/tests/unittests/test.pep.xml': Cannot get precursor information - no experiment given> occurred 6 times
```

**Comment by hroest** [comment]
I dont think this warning is new, I think this was present even before. It
just happens when you give it an empty experiment, see line 578

**Comment by grosenberger** [comment]
I spent now more time with the patch and tested the new options using the python bindings. I also checked the downstream TPP compatibility (4.8.0) and everything worked nicely.

:thumbsup:

**Comment by hendrikweisser** [comment]
Thanks for testing, George!

> so should we merge as is or test against TPP and remove the flag?

I'm now clearly in favour of removing the flag and the deprecated code.

**Comment by grosenberger** [comment]
Can we merge this feature?

File: src/openms/include/OpenMS/METADATA/PeptideHit.h [hunk 3]
```diff
@@ -84,6 +84,24 @@ namespace OpenMS
     };
     //@}
 
+    /// Analysis Result (containing search engine / prophet results)
+    class OPENMS_DLLAPI AnalysisResult
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/METADATA/PeptideHit.h]: I think the name could be better. Maybe just Scores and instead of analysis_type score_type

File: src/openms/include/OpenMS/METADATA/PeptideHit.h [hunk 4]
```diff
@@ -150,12 +177,16 @@ namespace OpenMS
 
     /// extracts the set of non-empty protein accessions from peptide evidences
     std::set<String> extractProteinAccessions() const;
+
 protected:
     AASequence sequence_;
 
     /// the score of the peptide hit
     double score_;
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/METADATA/PeptideHit.h]: then we should probably remove this score

File: src/openms/include/OpenMS/METADATA/PeptideIdentification.h [hunk 5]
```diff
@@ -159,11 +164,13 @@ namespace OpenMS
 
 
 protected:
+
     String id_; ///< Identifier by which ProteinIdentification and PeptideIdentification are matched
     std::vector<PeptideHit> hits_; ///< A list containing the peptide hits
     double significance_threshold_; ///< the peptide significance threshold
     String score_type_; ///< The score type (Mascot, Sequest, e-value, p-value)
     bool higher_score_better_; ///< The score orientation
+    String experiment_label_; ///< Identifier for the experiment in which this PeptideIdentification was found
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/METADATA/PeptideIdentification.h]: this should not be part of PeptideHit. Please move to metavalue

File: src/openms/source/METADATA/PeptideIdentification.cpp [hunk 6]
```diff
@@ -52,6 +52,7 @@ namespace OpenMS
     significance_threshold_(0.0),
     score_type_(),
     higher_score_better_(true),
+    experiment_label_(),
```
Comment (timosachsenberg) [inline src/openms/source/METADATA/PeptideIdentification.cpp]: see above

**Comment by timosachsenberg** [comment]
I see there is a need to encode the information in our PeptideHit / PeptideIdentification structure. As you noticed I am against extending these structures by e.g. a member for the experiment label. In my opinion this needs to be an optional meta value.
Let me use this example to point out a similar issue.in PeptideHit (and many other classes) e.g. the rank is encoding as a member - a property that should not be stored in the actual PeptideHit.
If this information needs to be passed down the object hierachy for whatever reason (e.g. the PeptideIdentifiation object is not available any more on writing the idXML), it is more appropriate to just annotate the object using a meta value. Otherwise we are more prone to inconsistencies and redundancy leading to larger core structures.

File: src/openms/include/OpenMS/METADATA/PeptideHit.h (hunk 3 above)
Comment (hroest) [inline src/openms/include/OpenMS/METADATA/PeptideHit.h]: well, in pepxml this is called analysis_result so I think the name makes sense, leading to less confusion than having 2 names for the same thing

**Comment by hroest** [comment]
@timosachsenberg I see your point and have made an according change. But in this case one more String is probably negligible in PeptideIdentification compared to a vector of PeptideHit (each having a AASequence, a vector of AnalysisResult and a vector of PeptideEvidence (each PE has an empty String, two ints and a char) ). So overall I understand your point but to me it seems the whole object is already quite bloated

glad you agree about PeptideHit, I think this only makes sense and is the only way to represent the data sensibly

**Comment by timosachsenberg** [comment]
Hi @hendrikweisser I meant that I see the need to store that information for pepXML but I don't want the structure extended (this includes PeptideHit).
Putting it a bit extreme: Turning PeptideHit and PeptideIdentification from a suboptimal state into a complete mess just to properly support pepXML is in my view not the right way to do it.

I see two alternatives:
- either carry the information along parallel to our structures (if possible - won't work between tools)
- encode the information in meta values

If you store it in meta information then I would suppose to not put the setter and getter into PeptiideHit but into a PepXMLHelper class that puts/extracts the analysis_result.

If we really need to include changes to peptidehit and peptideidentification we will orient on the mzIdentML requirements. Maybe they are that similar that I can be convinced ;)

**Comment by hendrikweisser** [comment]
@timosachsenberg:
> Hi @hendrikweisser

I think you meant @hroest there.

**Comment by timosachsenberg** [comment]
hehe ok

**Comment by hroest** [comment]
@timosachsenberg @grosenberger  I dont really see how this can be done without changing the PeptideIdentification. As far as I understand, currently the main usecase for this would be to write correct pepXML from Python.  I still think that TPP compatibility is important and here it seems to be needed since interProphet and Proteinprophet read in this information. 

Maybe @grosenberger can elaborate whether this is actually used a lot in the Aebersold lab at the moment.

**Comment by grosenberger** [comment]
Yes, we use this feature with pyOpenMS to write pepXML for downstream use with iProphet and ProteinProphet. Since 4.8.0, several new fields were introduced. I think the changes to pepXML export are very important to interface newer versions of TPP. AFAIK, this is the main use case for export of pepXML from OpenMS, so it would be great to have better compatibility.

**Comment by timosachsenberg** [comment]
I totally agree but the primary use case for PeptideIdentification and PeptideHit is not providing a storage model for PepXML/idXML/mzID but a data model for spectrum Identifications and PSMs.
I don't think it is a good idea to mix these two worlds - given that many of our data structures already present storage models for our own file formats which is not ideal in my view.
Isn't there a way we can go without moving PeptideIdentifications to a "hybrid storage model" of PepXML and idXML and mzIdentML? I just think that in the long run this might hurt and we might regret having taking this in the current model later.

**Comment by hendrikweisser** [comment]
I don't fully agree with Timo's point about "data model" vs. "storage model", but I don't understand why this is a contested issue, either. If there are fields in pepXML that are specific to pepXML and aren't used for anything internally, just stick them in meta values and have only the pepXML reading/writing code deal with them. So here I agree with Timo.
Hannes, where are meta values not sufficient ("I dont really see how this can be done without changing the PeptideIdentification.")?

**Comment by hroest** [comment]
well of course it is not impossible, but very annoying. Imagine that we have to basically store the following structure  in meta values

```
+          // <analysis_result analysis="peptideprophet">
+          //   <peptideprophet_result probability="0.0660" all_ntt_prob="(0.0000,0.0000,0.0660)">
+          //     <search_score_summary>
+          //       <parameter name="fval" value="0.7114"/>
+          //       <parameter name="ntt" value="2"/>
+          //       <parameter name="nmc" value="0"/>
+          //       <parameter name="massd" value="-0.027"/>
+          //       <parameter name="isomassd" value="0"/>
+          //     </search_score_summary>
+          //   </peptideprophet_result>
+          // </analysis_result>
```

so we will have to have 
- meta_value: "nr_analysis_results" => 1
- meta_value: "analysis_result_1_score_type" => peptideprophet
- meta_value: "analysis_result_1_score" => 0.0660
- meta_value: "analysis_result_1_score_nr" => 5
- meta_value: "analysis_result_1_score_1_value" => 0.7114
- meta_value: "analysis_result_1_score_1_name" => "fval"
- meta_value: "analysis_result_1_score_2_value" => 2
- meta_value: "analysis_result_1_score_2_name" => "ntt"
- etc...

which to me seems quite annoying and hard to parse. On the other hand, if we simply add a `std::vector<AnalysisResult> analysis_results_` member to PeptideHit, this does not have huge overhead but allows us to store the full datastructure. I am not quite sure how much overhead this incurs as it seems to differ depending on implementation  (see http://stackoverflow.com/questions/557997/what-is-the-overhead-cost-of-an-empty-vector) but it is probably on the order of 3 ptrs plus some depending on the platform (so maybe 24-30 bytes). We could save some space by using a `std::vector<AnalysisResult> * analysis_results_` which will have constant overhead of 8 bytes if it is NULL which seems quite defensible to me. 

Currently, the size of an empty PeptideHit is 104 bytes on my system, adding a `std::vector<AnalysisResult>*` increases it to 112 bytes while adding a `std::vector<AnalysisResult>` increases to size to 128 bytes (I guess there are 3 ptr needed here). As soon as we add any MetaValue, we pay 48 bytes for an empty MetaInfo object. In most real world applications, AASequence would not be empty and there would be n PeptideEvidence in the vector, increasing this by n*(24 + whatever std::string uses inside PeptideEvidence and internal AASequence structures) to maybe 128 + 10 [for accession string in PeptideEvidence] + 64 [for ptr vector of size 8 in AASequence] = 202 bytes or 250-274 bytes if there is a meta value.  This is on a 64 bit linux system, measured with `sizeof(x)`.
I would say using a pointer to `std::vector<AnalysisResult>` may be a good compromise for us all, only modestly increasing the empty size from 104 bytes to 112 bytes but making our lifes quite a bit easier. What do you say to this compromise? I think for a full PeptideHit, this would increase memory usage at most by 3-4% (8 bytes out of 200 to 250).

The ` PeptideIdentification::setExperimentLabel` is already implemented as a meta value, so that should be zero overhead if it is not present.


PS: for fun, I checked where the 104 bytes come from:
- 8 bytes vtable because of `virtual ~PeptideHit();`
- 8 bytes MetaInfoInterface (NULL ptr to MetaInfo)
- 48 bytes AASequence (this is an empty one)
- 24 bytes for the PeptideEvidence vector (this is an empty vector) 
- 16 bytes for UInt, Int, double


note that to this we have to add 
- 48 bytes for an empty MetaInfo object [assuming there is at least 1 metavalue which cost again extra, probably around 24 bytes / value]
- 8 bytes per Residue for the Residue ptr vector in AASequence
- n * 24 for the n PeptideEvidence objects
- 1 byte per char for the accession string in each PeptideEvidence object

giving probably a total of `104 + y * 48 + x * 8 + n * (24 + z) + m * 24` if I counted correctly 

where 
- y = [0,1] whether we have MetaValues
- x = number of Residues in AASequence
- z = length of ProteinAccession string
- n = number of PeptideEvidence objects
- m = number of Meta Values

This simply accounts for the empty STL objects and the size of each individual object but does not count per-element-overhead that structures such as `std::map` have (as in MetaInfo). We thus expect the minimal size to be 104 but in reasonable applications to be closer to 200 - 300 bytes with one PeptideEvidence and zero to two MetaValues.

note that AASequence is composed of:
- 8 bytes vtable 
- 24 bytes for empty Residue* vector  
- 8 bytes for ResidueModification* Nterm
- 8 bytes for ResidueModification* Cterm
- n * 8 bytes for each Residue (on average 7 Residues)

note that MetaInfo uses per value:
- 4 bytes for UInt index
- DataValue of size about 20 bytes
  - 4 bytes for DataType enum
  - 8 bytes for union data_ 
  - 8 bytes for std::string for Unit

**Comment by timosachsenberg** [comment]
Hi @hroest thanks for the detailed analysis.
My primary concern is that semantics of the plain objects are blured if to much is added as members. On the other hand I completly understand that introducing a complete separation is too much work without immidiate benefits.
If we commit ourself to providing a storage model for different data formats I think providing members like e.g.: PepXMLAnalysisResult* analysis_results; (as you bascially proposed for the vector*), could provide a cleaner separation of concepts.
Also this would allow to move meta values introduced like in PeptideIdentification::setExperimentLabel (which imho are not neccessarly a base property of a spectrum identification object) to the data format specific members (e.g. PepXMLAnalysisResult* or similar).
Of course there might be overlap between the data formats and figuring out what is specific for a data format or common to all might be some work. Then again, I think this is not a big problem as this could be done incrementally.

**Comment by hendrikweisser** [comment]
> Imagine that we have to basically store the following structure in meta values

How general or specific does that have to be, though? If it can be specific to a PeptideProphet analysis that always has the same parameters, you could just store it in five (or six) meta values: `"peptideprophet_fval": 0.7114` etc. (the PP probability gets stored as the score of the PeptideHit, I think).

Alternatively, since these are auxiliary scores for peptide hits, maybe this would be the point where we should introduce multiple scores into the data model (with one marked as the "main score" for backward compatibility)? This has to happen anyway.

**Comment by hroest** [comment]
@timosachsenberg indeed, I think your suggestion makes sense, except I still would not call it PepXMLAnalysisResult because it may be that we can re-purpose the same datastructure for another file format in which case it might not make sense to have it named for PepXML only. However, we can currrently remark in the documentation that this is mainly a PepXML structure. In general, I think we always have the problem that we try to have general datastructures that can map to all underlying file formats but are specific enough not to loose information. I dont think that we can ever fully satisfy both constraints of being complete and being lightweight. one could try to have some elaborate type hierarchy where we have a BaseEverything, e.g. a BasePeptideEvidence, BasePeptideHit, BaseXXX and then a specific implementation for every file format but I dont really see that as being easier conceptually. Also, in the end, there are not thousands of file formats but currently 3 for identification. Alternatively, one could have one of everything with ptrs to specific substructures, e.g. in the worst case

```
class PeptideHit
{

private:
/// some shared stuff
AASequence sequence_;
/// some extension
PepXMLPeptideHit* pep_xml_hit_;
MzIdentMLPeptideHit* mz_ident_hit_
IdXMLPeptideHit* id_xml_hit_;
}
```
In general this is more expensive in memory than inheritance (which costs us 1 vtable ptr) and also somewhat more cumbersome is you always have to check if x == NULL then get data. On the other hand, it is currently an easy solution to our problem and allows us to use the same type all over the code, so that gets easier.

One thing I really like is your idea of packing all PepXML stuff into one structure - at first. On the other hand (i), conceptually, the ExperimentLabel really does not belong into the AnalysisResult structure. Secondly (ii) the ExperimentLabel should be at the level of PeptideIdentification and not on PeptideHitLevel. Putting it there would mean we (iii) we cannot store an ExperimentLabel in an PeptideIdentification if it does not have at least one PeptideHit and (iv) we could run into issues that several PeptideHit objects have different labels. So I am a bit torn ... but if we would do it, then it would look like this:

```
class PepXMLPeptideHit
{
private:
std::vector<AnalysisResult> analysis_results_;
String experiment_label_;
}
```

**Comment by timosachsenberg** [comment]
Hi @hroest 
Putting the experiment_label in the PeptideHit was just an example how this might result in cleaner code. It is probably better put at the level of PeptideIdentification.
PepXMLAnalysisResult is also just name that I quickly made up because AnalysisResult was a bit to general in my view (e.g. MzIdentML has AnalysisData, AnalysisSoftware, ...).
I saw similar constructs in the past that used the Suffix "Detail" to mark such kind of object extensions e.g. something like PepXMLPSMDetail or a more general prefix than "PepXML").  
I also agree that introducing a class hierarchy and use polymorphism is too involved.
So what is the general opinion on this - could we agree on putting the specific AnalysisResults in a class ? @hendrikweisser, what do you think. I think we will need to model multiple scores but this use case is a list of a set of multiple scores. Which means that even with multiple scores it would require some extra tagging/naming to keep the hierarchy. If we allow multiple scores (which I think we should) then I would like a flat list of scores, ideally annotated with CV term and how they are compared (higher score better).

**Comment by hroest** [comment]
> Alternatively, since these are auxiliary scores for peptide hits, maybe this would be the point where we should introduce multiple scores into the data model (with one marked as the "main score" for backward compatibility)? This has to happen anyway.

@hendrikweisser This is exactly what this PR is doing, or did I not understand you correctly? 

>  If we allow multiple scores (which I think we should) then I would like a flat list of scores, ideally annotated with CV term and how they are compared (higher score better).

@timosachsenberg it is always possible to model the "easy" case of a simple list of flat scores with our approach. However the approach  here is more powerful.

@timosachsenberg yes, that sounds great, I like the "Detail" idea. It is similar to what people do in database designs if a table becomes too unhandy or carries too much information to still be useful for most queries. Right now, the current design allows us, I think, to do what you want: it is a hierarchical system that allows you to store multiple `AnalysisResult` objects, each of which knows its main score, whether higher is better and has information about all the subscores used to generate this result. So it is possible for example to have a single peptide hit that carries peptideprophet scores from Mascot and another result for the same peptide with a percolator score where the user combined mascot and sequest scores (for example, not saying that this is what you should do, but it is possible). For both scores peptideprophet and percolator we would retian the information whether higher score is better and which subscores were used to generate the score. Does this sound like what you had in mind? 

```
String score_type; // e.g. peptideprophet / interprophet
bool higher_is_better; // is higher score better ?
double main_score; // posterior probability for example
std::map<String, double> sub_scores; /// additional scores attached to the original, aggregated score
```
if you like, we could add another vector, e.g. `std::map<String, bool>` to store higher=better for each subscore, but I dont think this is ever necessary.


note that mzIdentML at the moment seems to store the score information in cv terms, e.g. 
```
<cvParam accession="MS:1001171" name="Mascot:score" cvRef="PSI-MS" value="62.72"/>
<cvParam accession="MS:1001172" name="Mascot:expectation value" cvRef="PSI-MS" value="0.000863428353297988"/>
```

so in mzIdentML it seems like we can store Mascot and Sequest and all other types of scores, albeit in a flat hierarchy only.

**Comment by hendrikweisser** [comment]
My opinion is that we should just model the structure that is used in mzIdentML. This would have the following advantages:
- easy compatibility to mzIdentML, i.e. the main ID file format
- flat hierarchy makes it easier to find specific scores
- score orientation is stored only once, with the score type
- "main score" is ill-defined, may depend on the application (not for PeptideProphet, but for Mascot, Percolator, MS-GF+, ...)

In the mzIdentML structure, the hierarchical grouping is implicitly given by the CV terms. There are pros and cons for a flat or a hierarchical implementation, but in case of doubt I would prefer to go with the way it's done in mzIdentML. (I'm pretty sure the pepXML scoring case is covered there, too.)

**Comment by hroest** [comment]
I dont think that we can easily translate everything we find in pepxml into nice CV terms that can also be found in mzIdentML simply because, basically, anybody can write whatever they like into pepxml. People just recently extended pepxml for SWATH data and they just added a bunch of new scores. As this is not documented and no CV terms exist, I dont think we can represent the hierarchy using the same approach as in mzIdentML. So I dont quite see where that leaves us but I believe that 

@uweschmitt pyOpenMS seems to work with the latest fix

```
>>> import pyopenms
>>> pp = pyopenms.PeptideHit_AnalysisResult()
>>> a = pyopenms.String("A")
>>> pp.sub_scores = {  a: 8}
>>> pp.sub_scores = { "a" : 8}
Segmentation fault
```

the segfault is not so great of course but more an autowrap issue. It seems we are doing something like this

```
|        def __set__(self, dict sub_scores):
-            cdef libcpp_map[_String, double] * v0 = new libcpp_map[_String, double]()
3            for key, value in sub_scores.items():
3               deref(v0)[ deref(<_String *> (<String> key).inst.get()) ] = <double> value
3            self.inst.get().sub_scores = deref(v0)
3            del v0

```

maybe before iterating, some sort of check would make sense? e.g. `assert isinstance(key, String)` - anyways conversion between bytes and OpenMS string is not always as great as it could be ...

**Comment by timosachsenberg** [comment]
+1
